from snakemake.utils import min_version
min_version("6.0")


rule bcftools_norm:
    input:
        ref = "genome.fasta",
        vcf = "variants.vcf"
    output:
        vcf = "variants.norm.vcf"
    conda:
        "bcftools.yaml"
    log:
        "bcftools.norm.log"
    shell:
        """
        bcftools norm \
            --fasta-ref {input.ref} \
            --check-ref s \
            --multiallelics -both \
            -o {output.vcf} \
            {input.vcf} \
            > {log}
        """


rule bcftools_norm_noref:
    input:
        vcf = "variants.vcf"
    output:
        vcf = "variants.norm.vcf"
    conda:
        "bcftools.yaml"
    log:
        "bcftools.norm.noref.log"
    shell:
        """
        bcftools norm \
            --multiallelics -both \
            -o {output.vcf} \
            {input.vcf} \
            > {log}
        """


rule bcftools_sort:
    input:
        vcf = "variants.vcf"
    output:
        vcf = "variants.sorted.vcf.gz"
    conda:
        "bcftools.yaml"
    log:
        "bcftools.sort.log"
    shell:
        """
            bcftools sort \
                -o {output.vcf} \
                -O z \
                {input.vcf}
        """


rule bcftools_index:
    input:
        vcf = "variants.sorted.vcf.gz"
    output:
        tbi = "variants.sorted.vcf.gz.tbi"
    conda:
        "bcftools.yaml"
    shell:
        """
            bcftools index -t {input.vcf}
        """